core/src/main/java/io/parsingdata/metal/expression/value/BinaryValueExpression.java

```diff
@@ -16,35 +16,48 @@
 
 package io.parsingdata.metal.expression.value;
 
-import static io.parsingdata.metal.Util.checkNotNull;
-
 import io.parsingdata.metal.data.Environment;
+import io.parsingdata.metal.data.OptionalValueList;
 import io.parsingdata.metal.encoding.Encoding;
 
+import static io.parsingdata.metal.Util.checkNotNull;
+
 public abstract class BinaryValueExpression implements ValueExpression {
 
-    private final ValueExpression _lop;
-    private final ValueExpression _rop;
+    private final ValueExpression lop;
+    private final ValueExpression rop;
 
     public BinaryValueExpression(final ValueExpression lop, final ValueExpression rop) {
-        _lop = checkNotNull(lop, "lop");
-        _rop = checkNotNull(rop, "rop");
+        this.lop = checkNotNull(lop, "lop");
+        this.rop = checkNotNull(rop, "rop");
     }
 
     @Override
-    public OptionalValue eval(final Environment env, final Encoding enc) {
-        final OptionalValue lv = _lop.eval(env, enc);
-        if (!lv.isPresent()) { return lv; }
-        final OptionalValue rv = _rop.eval(env, enc);
-        if (!rv.isPresent()) { return rv; }
-        return eval(lv.get(), rv.get(), env, enc);
+    public OptionalValueList eval(final Environment env, final Encoding enc) {
+        return evalLists(lop.eval(env, enc), rop.eval(env, enc), env, enc);
+    }
+
+    private OptionalValueList evalLists(final OptionalValueList lvl, final OptionalValueList rvl, final Environment env, final Encoding enc) {
+        if (lvl.isEmpty()) { return makeListWithEmpty(rvl.size); }
+        if (rvl.isEmpty()) { return makeListWithEmpty(lvl.size); }
```

Comment:
These two lines of codes create behavior that should be documented *somewhere* in the Javadoc of this class (either on the class level Javadoc or on the public `eval` above).